src/Mvc/Mvc.ViewFeatures/src/Infrastructure/DefaultTempDataSerializer.cs

```diff
@@ -73,6 +81,53 @@ internal class DefaultTempDataSerializer : TempDataSerializer
             return deserialized;
         }
 
+        private static object DeserializeArray(in JsonElement arrayElement)
+        {
+            if (arrayElement.GetArrayLength() == 0)
+            {
+                // We have to infer the type of the array by inspecting it's elements.
+                // If there's nothing to inspect, return a null value since we do not know
+                // what type the user code is expecting.
+                return null;
+            }
+
+            if (arrayElement[0].Type == JsonValueType.String)
+            {
+                var array = new List<string>();
+
+                foreach (var item in arrayElement.EnumerateArray())
+                {
+                    array.Add(item.GetString());
+                }
+
+                return array.ToArray();
+            }
+            else if (arrayElement[0].Type == JsonValueType.Number)
+            {
+                var array = new List<int>();
+
+                foreach (var item in arrayElement.EnumerateArray())
+                {
+                    array.Add(item.GetInt32());
```

Comment:
What if the elements are actually strings. If that string for the first element will be a number, this will eventually fail. Is there a model, where the user can specify explicitly, what type of data to expect or just enforce `string` array somehow?

Reply:
Not super sure I follow. Is this if the collection is heterogeneous? We control what goes in here so we only allow collection of strings or collection of ints.